File: C1/PDOO/Material Complementario/Tema7/VillanosRB-ClaseTeoria/lib/villano.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Villanos
  class Villano
    @@Asociacion = "VillanosSA"

    def initiliaze(n,e,f)
      @nombre = n
      @escondite_secreto = e
      @fortuna = f
      @planes = Array.new
      @mayordomo = nil
    end

    def set_fortuna(cantidad)

    end

    def seguir_plan_prioritario()
      terminado = false
      plan = primer_plan_activo
      if plan != nil
        puedo_continuar = true
        while !terminado and puedo_continuar do
          terminado = plan.get_terminado
          gasto = plan.inversion_paso_siguiente
          puedo_continuar = gasto <= @fortuna
          if puedo_continuar
            ganancia = plan.dar_paso_siguiente
            set_fortuna(@fortuna-gasto+ganancia)
          end
        end
      end
      terminado = plan != nil and terminado
      return terminado
    end

    def primer_plan_activo()

    end

    def self.set_asociacion(asociacion)

    end
    private :set_fortuna

    # si el set_asociacion fuera privado
    # private_class_method :set_asociacion
  end
end

File: C1/PDOO/Material Complementario/Tema7/Villanos-ClaseTeoria/src/villanos/Villano.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package villanos;

import java.util.ArrayList;
import esbirros.Mayordomo;

/**
 *
 * @author salva
 */
public class Villano {

    //atributos de instancia
    public String nombre;
    String esconditeSecreto;
    private int fortuna;

    //atributos de clase
    public static String Asociacion = "VillanosSA";

    //atributos de referencia
    private ArrayList<Plan> planes;
    Mayordomo mayordomo;
    
   // boolean puedoContinuar;

    //constructor por defecto
    public Villano(String nombre, String esconditeSecreto, int fortuna) {
        this.nombre = nombre;
        this.esconditeSecreto = esconditeSecreto;
        this.fortuna = fortuna;
        this.mayordomo = null;
        this.planes = new ArrayList<>();
        planes.add(new Plan());
        //this.puedoContinuar = null;
    }

    private void setFortuna(int cantidad) {

    }

    public boolean seguirPlanPrioritario() {
        boolean terminado = false;
        Plan plan = primerPlanActivo();
        if (plan != null){
            boolean puedoContinuar = true;
            while (!terminado && puedoContinuar){
                terminado = plan.getTerminado();
                int gasto = plan.inversionPasoSiguiente();
                puedoContinuar = gasto <= fortuna;
                if (puedoContinuar){
                    int ganancia = plan.darPasoSiguiente();
                    this.setFortuna(fortuna-gasto+ganancia);
                }
            }
        }
        terminado = plan != null && terminado;
        return terminado;
    }

    public Plan primerPlanActivo() {

    }

    public static void setAsociacion(Asociacion Asociacion) {

    }
}
